### ISSUE TITLE
OLMo-2 1B `grpo_fast.py` strange behaviour around 2.3k steps

### ISSUE THREAD

**Issue by BrownianNotion** [allenai/open-instruct#1610]
Hey team,
I have been running stage 1 of OLMo-2 1B RLVR and have encountered strange behaviour at around 2.3k steps. 

#### Issue
Metrics including grad norm, loss (both kl and policy), vllm vs local log probs,  spike experience large sustained fluctuations. This happens for both the original settings, and on newer settings (truncated importance sampling + clip higher + inflight updates). I have included the wandb metrics below - the new settings are in grey, the original in pink. 

**Note:** I performed training on a single node, unlike the original, which was done on 2 nodes. I have attempted to adjust the batch size related settings to maintain equivalence.

<img width="1683" height="812" alt="Image" src="https://github.com/user-attachments/assets/6f2256e7-1a9c-4951-a9bf-bb1ac966dd6f" />

<img width="1855" height="781" alt="Image" src="https://github.com/user-attachments/assets/3f99e91e-c88b-4ae3-9e03-26f8a6d9c086" />

<img width="1864" height="776" alt="Image" src="https://github.com/user-attachments/assets/7ba1eecc-5b56-4763-b183-bc708621e13d" />

Has anyone experienced anything similar? The evaluations across the standard [benchmarks](https://github.com/allenai/olmes/blob/main/oe_eval/configs/task_suites.py#L671-L683) don't seem to show any degradation (see below, used 512 max gen tokens) and doing some quick vibe checks (eg. 2+2, gsm8k like problems) doesn't show any strange behaviour. Haven't done too much digging yet as a result, thought I would ask if anyone has seen the same thing first.

| Task       |   OLMo-2-1B-RLVR1-tis-dapo-inflight-updates (grey wandb run) |   OLMo-2-0425-1B-RLVR1 (hugging-face checkpoint) |
|:-----------|------------------------------------------:|-----------------------:|
| BBH        |                                      33.8 |                   33.6 |
| DROP       |                                      29   |                   29.8 |
| GSM8K      |                                      71.5 |                   68.6 |
| HumanEval  |                                      40.4 |                   43.3 |
| IFEval     |                                      68.8 |                   66.9 |
| MATH       |                                      17.6 |                   15.6 |
| MMLU       |                                      38.2 |                   39   |
| PopQA      |                                      11.7 |                   12.8 |
| TruthfulQA |                                      49   |                   48.3 |
| Average    |                                      40   |                   39.8 |

#### Run commands
Original settings (my attempt at reconciliation, apologies in advance for errors).
```bash
python grpo_fast.py \
    --exp_name olmo2_1b_rlvr1_updated_open_instruct \
    --seed 1 \
    --num_epochs 1 \
    --total_episodes 2000000 \
    --learning_rate 5e-7 \
    --lr_scheduler_type constant \
    --per_device_train_batch_size 1 \
    --beta 0.01 \
    --kl_estimator 3 \
    --num_mini_batches 4 \
    --num_samples_per_prompt_rollout 16 \
    --num_unique_prompts_rollout 48 \
    --model_name_or_path allenai/OLMo-2-0425-1B-DPO \
    --model_revision main \
    --tokenizer_name_or_path allenai/OLMo-2-0425-1B-DPO \
    --tokenizer_revision main \
    --add_bos \
    --chat_template_name tulu \
    --dataset_mixer_list allenai/RLVR-GSM-MATH-IF-Mixed-Constraints 1.0 \
    --dataset_mixer_list_splits train \
    --dataset_mixer_eval_list allenai/RLVR-GSM-MATH-IF-Mixed-Constraints 16 \
    --dataset_mixer_eval_list_splits train \
    --dataset_cache_mode local \
    --max_prompt_token_length 2048 \
    --response_length 2048 \
    --pack_length 4096 \
    --stop_strings "</answer>" \
    --temperature 1.0 \
    --apply_verifiable_reward true \
    --remap_verifier ifeval=ifeval_old \
    --non_stop_penalty \
    --non_stop_penalty_value 0.0 \
    --clip_higher 0.2 \
    --truncated_importance_sampling_ratio_cap 0.0 \
    --inflight_updates False \
    --async_steps 1 \
    --advantage_normalization_type standard \
    --deepspeed_stage 2 \
    --deepspeed_zpg 6 \
    --fused_optimizer True \
    --gradient_checkpointing \
    --num_learners_per_node 6 \
    --vllm_num_engines 2 \
    --vllm_tensor_parallel_size 1 \
    --vllm_gpu_memory_utilization 0.9 \
    --vllm_sync_backend nccl \
    --local_eval_every 100 \
    --save_freq 200 \
    --save_traces True \
    --checkpoint_state_freq 50 \
    --checkpoint_state_dir output/olmo2_1b_rlvr1_ds_state_checkpoints \
    --keep_last_n_checkpoints 3 \
    --output_dir output/olmo2_1b_rlvr1_hf_checkpoints \
    --with_tracking
```
Newer settings (truncated importance sampling_cap=2.0 + clip_higher=0.27 + inflight_updates=True)

```bash
python grpo_fast.py \
    --exp_name olmo2_1b_rlvr_stage1 \
    --seed 1 \
    --num_epochs 1 \
    --total_episodes 2000000 \
    --learning_rate 5e-7 \
    --lr_scheduler_type constant \
    --per_device_train_batch_size 1 \
    --beta 0.01 \
    --kl_estimator 3 \
    --num_mini_batches 4 \
    --num_samples_per_prompt_rollout 16 \
    --num_unique_prompts_rollout 48 \
    --model_name_or_path allenai/OLMo-2-0425-1B-DPO \
    --model_revision main \
    --tokenizer_name_or_path allenai/OLMo-2-0425-1B-DPO \
    --tokenizer_revision main \
    --add_bos \
    --chat_template_name tulu \
    --dataset_mixer_list allenai/RLVR-GSM-MATH-IF-Mixed-Constraints 1.0 \
    --dataset_mixer_list_splits train \
    --dataset_mixer_eval_list allenai/RLVR-GSM-MATH-IF-Mixed-Constraints 16 \
    --dataset_mixer_eval_list_splits train \
    --dataset_cache_mode local \
    --max_prompt_token_length 2048 \
    --response_length 2048 \
    --pack_length 4096 \
    --stop_strings "</answer>" \
    --temperature 1.0 \
    --apply_verifiable_reward true \
    --remap_verifier ifeval=ifeval_old \
    --non_stop_penalty \
    --non_stop_penalty_value 0.0 \
    --deepspeed_stage 2 \
    --deepspeed_zpg 6 \
    --fused_optimizer True \
    --gradient_checkpointing \
    --num_learners_per_node 6 \
    --vllm_num_engines 2 \
    --vllm_tensor_parallel_size 1 \
    --vllm_gpu_memory_utilization 0.9 \
    --vllm_sync_backend nccl \
    --local_eval_every 100 \
    --save_freq 200 \
    --save_traces True \
    --checkpoint_state_freq 50 \
    --checkpoint_state_dir output/olmo2_1b_rlvr_stage1_ds_state_checkpoints \
    --keep_last_n_checkpoints 3 \
    --output_dir output/olmo2_1b_rlvr_stage1_hf_checkpoints \
    --with_tracking
```

**Comment by hamishivi** [comment]
@BrownianNotion Interesting! I think generally if downstream and eval metrics are still going up you are fine, and seeing KL go up a lot later in training can happen... it could indicate that the model has improved enough such that further improvements require larger changes, or be more unstable training. One thing I would look at is the 'real batch size' - is this going down over training? As the model gets better, it completely solves more instances, and so the 'effectiv batch size' (samples with non-zero advantages) goes down, and the training becomes more unstable.

One thing I would recommend is to turn on active sampling w/ multiple async steps. Active sampling ensures that we only build a batch of samples where at least one sample was different to the others, at the cost of throwing out some samples (so can take longer between steps). This can be useful to turn on with no_resample_pass_rate 0.9 or similar, which just prevents resampling instances the model has solved (hopefully reducing some of the waste of active sampling). Overall, it makes sure that the real batch size doesnt degrade, and the model is trained on increasingly difficult tasks, which could help with reducing instabilities later in training like you see there (if that is indeed the issue). But it could also just be the model is just finding it more difficult to optimize to improve further!

**Comment by BrownianNotion** [comment]
@hamishivi Thanks for the tips, I'll give those settings a go on the next run! `batch/total_prompts` does indeed go down, but it follows a similar trajectory to the old code, which did not experience metric spikes. I'll keep an eye out anything strange but for now it should be ok with the similar evals. Thanks!

<img width="1856" height="497" alt="Image" src="https://github.com/user-attachments/assets/9fa05a38-c513-4c52-b1f9-dfbd5ee87b14" />

**Comment by hamishivi** [comment]
No worries, please re-open if there are still problems!
